### PR TITLE
feat: introduce `ChainSpecParser` generic in cli types

### LINKED ISSUE #10566: Integrate Chainspec parser in CLI
### Describe the feature

In order to support custom chainspec on the CLI level, which is usually supported by an argument "--chain mainnet" e.g. we'd need to support custom chainspec parsing, which was the issue that resulted in the revert of #9877

we already have:

https://github.com/paradigmxyz/reth/blob/516ae50320a1a8fecddd19b1685680fce325934c/crates/cli/cli/src/chainspec.rs#L10-L10

and https://github.com/paradigmxyz/reth/blob/516ae50320a1a8fecddd19b1685680fce325934c/crates/cli/cli/src/lib.rs#L24-L24

this issue is most likely an intermediary step to unblock #9877 which already separates OP and reth cli types

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by klkvr**
Closes https://github.com/paradigmxyz/reth/issues/10566

Adds `ChainSpecParser` generic to CLI types and uses it as parser for chainspec. For `Cli` and `NodeCommand` it has a default value of `DefaultChainSpecParser` which relies on current feature gated functions.

I've changed chainspec generic `ChainSpecParser` to an AT as this allows us to only introduce a single generic on cli types. Instead of `TypedValueParser` supertrait added a `parser` method which allows code reuse instead of duplicating clap parser impls. Also removed `possible_values` impl as it duplicated values we have in help message and didn't mention file option.

### CODE REVIEW

File: crates/cli/cli/src/chainspec.rs [hunk 1]
```diff
@@ -2,12 +2,40 @@ use std::sync::Arc;
 
 use clap::builder::TypedValueParser;
 
+#[derive(Debug, Clone)]
+struct Parser<C>(std::marker::PhantomData<C>);
+
+impl<C: ChainSpecParser> TypedValueParser for Parser<C> {
+    type Value = Arc<C::ChainSpec>;
```
Comment (mattsse) [inline crates/cli/cli/src/chainspec.rs:9]: this should also override possible_values

File: bin/reth/src/cli/mod.rs [hunk 2]
```diff
@@ -35,10 +34,10 @@ pub use crate::core::cli::*;
 /// This is the entrypoint to the executable.
 #[derive(Debug, Parser)]
 #[command(author, version = SHORT_VERSION, long_version = LONG_VERSION, about = "Reth", long_about = None)]
-pub struct Cli<Ext: clap::Args + fmt::Debug = NoArgs> {
+pub struct Cli<Ext: clap::Args + fmt::Debug = NoArgs, C: ChainSpecParser = DefaultChainSpecParser> {
```
Comment (mattsse) [inline bin/reth/src/cli/mod.rs:37]: I'd like to change the order here: C,Ext

**Review by mattsse — CHANGES_REQUESTED** [review]
some nits, 

wdyt @joshieDo ?

File: crates/cli/cli/src/chainspec.rs (hunk 1 above)
Comment (klkvr) [reply crates/cli/cli/src/chainspec.rs:9]: It causes duplicating data in help message (we also have values list in `chain_help()`) https://github.com/paradigmxyz/reth/actions/runs/10587216380/job/29337535172

**Review by joshieDo — APPROVED** [review]
should unlock the op bin, nice

**Review by mattsse — APPROVED** [review]
unblocks #9877

**Pull request merged**